sdk/search/azure-search-documents/azure/search/documents/_internal/_search_index_document_batching_client.py

```diff
@@ -0,0 +1,326 @@
+    ABC = abc.ABCMeta("ABC", (object,), {"__slots__": ()})  # type: ignore
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any, Union
+    from azure.core.credentials import AzureKeyCredential
+
+class PersistenceBase(ABC):
+    @abc.abstractmethod
+    def add_queued_actions(self, actions, **kwargs):
+        # type: (*str, List[IndexAction], dict) -> None
+        pass
+
+    @abc.abstractmethod
+    def add_succeeded_action(self, action, **kwargs):
+        # type: (*str, IndexAction, dict) -> None
+        pass
+
+    @abc.abstractmethod
+    def add_failed_action(self, action, **kwargs):
+        # type: (*str, IndexAction, dict) -> None
+        pass
+
+    @abc.abstractmethod
+    def remove_queued_action(self, action, **kwargs):
+        # type: (*str, IndexAction, dict) -> None
+        pass
+
+class SearchIndexDocumentBatchingClient(HeadersMixin):
+    """A client to do index document batching.
+
+    :param endpoint: The URL endpoint of an Azure search service
+    :type endpoint: str
+    :param index_name: The name of the index to connect to
+    :type index_name: str
+    :param credential: A credential to authorize search client requests
+    :type credential: ~azure.core.credentials.AzureKeyCredential
+    :keyword int window: how many seconds if there is no changes that triggers auto flush.
+        if window is less or equal than 0, it will disable auto flush
+    :keyword int batch_size: batch size. Default to 1000. It only takes affect when auto_flush is on
+    :keyword persistence: persistence hook. If it is set, the batch client will dump actions queue when it changes
+    :paramtype persistence: PersistenceBase
+    :keyword str api_version: The Search API version to use for requests.
+    """
+    # pylint: disable=too-many-instance-attributes
+    _ODATA_ACCEPT = "application/json;odata.metadata=none"  # type: str
+    _DEFAULT_WINDOW = 0
+    _DEFAULT_BATCH_SIZE = 1000
+
+    def __init__(self, endpoint, index_name, credential, **kwargs):
+        # type: (str, str, AzureKeyCredential, **Any) -> None
+
+        api_version = kwargs.pop('api_version', None)
+        validate_api_version(api_version)
+        self._batch_size = kwargs.pop('batch_size', self._DEFAULT_BATCH_SIZE)
+        self._window = kwargs.pop('window', self._DEFAULT_WINDOW)
+        self._auto_flush = self._window > 0
+        self._index_documents_batch = IndexDocumentsBatch()
+        self._endpoint = endpoint  # type: str
+        self._index_name = index_name  # type: str
+        self._index_key = None
+        self._credential = credential  # type: AzureKeyCredential
+        self._client = SearchIndexClient(
+            endpoint=endpoint, index_name=index_name, sdk_moniker=SDK_MONIKER, **kwargs
+        )  # type: SearchIndexClient
+        self._reset_timer()
+        self._persistence = kwargs.pop('persistence', None)
+
+    def cleanup(self, flush=True, raise_error=False):
```

Comment:
Is there a reason this function is public?
Outside of closing the client - in what other scenarios would someone call this?

Reply:
If auto_flush is on, the batching client will create a thread, we need this to shut down the thread.

Reply:
Yes but would you shut down the thread for any reason other than closing the client?